lib/msal-common/src/unifiedCache/UnifiedCacheManager.ts

```diff
@@ -61,47 +111,341 @@ export class UnifiedCacheManager {
+            if (!!target && CacheHelper.getCredentialType(key) != CredentialType.ID_TOKEN.toString().toLowerCase()) {
+                matches = matches && CacheHelper.matchTarget(key, target);
+            }
+
+            if (matches) {
+                matchingCredentials[key] = cacheCredentials[key];
+            }
+        });
+
+        return matchingCredentials;
     }
 
     /**
-     * append account to the in memory cache
+     * returns a boolean if the given account is removed
      * @param account
      */
-    addAccountEntity(account: AccountEntity): void {
-        const accKey = account.generateAccountKey();
-        if (!this.inMemoryCache.accounts[accKey]) {
-            this.inMemoryCache.accounts[accKey] = account;
+    removeAccount(account: AccountEntity): boolean {
+        const cache = this.getCacheInMemory();
+        const accountKey = account.generateAccountKey();
+
+        if (!!cache.accounts[accountKey]) {
+            delete cache.accounts[accountKey];
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * returns a boolean if the given account is removed
+     * @param account
+     */
+    removeAccountContext(account: AccountEntity): boolean {
+        const cache = this.getCacheInMemory();
+
+        const accountId = account.generateAccountId();
+
+        // TODO: Check how this should be done, do we just remove the account or also the associated credentials always?
+        Object.keys(cache.accessTokens).forEach((key) => {
+            if (
+                cache.accessTokens[key].generateAccountId() === accountId
+            )
+                this.removeCredential(cache.accessTokens[key]);
+        });
+
+        Object.keys(cache.idTokens).forEach((key) => {
+            if (cache.idTokens[key].generateAccountId() === accountId)
+                this.removeCredential(cache.idTokens[key]);
+        });
+
+        Object.keys(cache.idTokens).forEach((key) => {
+            if (cache.idTokens[key].generateAccountId() === accountId)
+                this.removeCredential(cache.idTokens[key]);
+        });
+
+        this.removeAccount(account);
+        return true;
+    }
+
+    /**
+     * returns a boolean if the given credential is removed
+     * @param credential
+     */
+    removeCredential(credential: Credential): boolean {
+        const cache = this.getCacheInMemory();
+
+        switch (credential.credentialType) {
+            case CredentialType.ID_TOKEN:
+                delete cache.idTokens[
+                    credential.generateCredentialKey()
+                ];
+                return true;
+            case CredentialType.ACCESS_TOKEN:
+                delete cache.accessTokens[
+                    credential.generateCredentialKey()
+                ];
+                return true;
+            case CredentialType.REFRESH_TOKEN:
+                delete cache.refreshTokens[
```

Comment:
Helper functions would be nice here, as this is also leaking implementation details.

Reply:
Done